### PR TITLE
CIRC-938 Delete scheduled hold shelf expiration notice for filled request

### PR DESCRIPTION
**Pull request by alexanderkurash**
Resolves [CIRC-938](https://issues.folio.org/browse/CIRC-938).

## Purpose
Delete hold shelf expiration notice for a request that is
closed as filled.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java [hunk 1]
```diff
@@ -61,9 +68,32 @@ private RequestScheduledNoticeHandler(RequestRepository requestRepository,
...
+          .thenCompose(res -> res.after(records -> sendNotice(records, notice)))
+          .thenCompose(res -> res.after(records -> updateNotice(records, notice)));
+      }));
+  }
+
+  private CompletableFuture<Result<Request>> deleteNoticeIfRequestIsClosedAsFilled(
+    Request request, ScheduledNotice notice) {
+
+    if (notice.getTriggeringEvent().equals(HOLD_EXPIRATION) &&
+      request.getStatus().equals(RequestStatus.CLOSED_FILLED)) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java:87]: Check for NPE

File: src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java [hunk 2]
```diff
@@ -61,9 +68,32 @@ private RequestScheduledNoticeHandler(RequestRepository requestRepository,
 
   private CompletableFuture<Result<ScheduledNotice>> handleRequestNotice(ScheduledNotice notice) {
     return requestRepository.getById(notice.getRequestId())
-      .thenApply(r -> r.map(RequestAndRelatedRecords::new))
-      .thenCompose(r -> r.after(records -> sendNotice(records, notice)))
-      .thenCompose(r -> r.after(records -> updateNotice(records, notice)));
+      .thenCompose(r -> r.after(request -> deleteNoticeIfRequestIsClosedAsFilled(request, notice)))
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java:71]: Use afterWhen?

File: src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java [hunk 3]
```diff
@@ -61,9 +68,32 @@ private RequestScheduledNoticeHandler(RequestRepository requestRepository,
...
+      .thenCompose(r -> r.after(request -> {
+        if (request == null) {
+          return completedFuture(succeeded(notice));
+        }
+
+        return completedFuture(succeeded(new RequestAndRelatedRecords(request)))
+          .thenCompose(res -> res.after(records -> sendNotice(records, notice)))
+          .thenCompose(res -> res.after(records -> updateNotice(records, notice)));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java:79]: Extract method

File: src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java [hunk 4]
```diff
@@ -61,9 +68,33 @@ private RequestScheduledNoticeHandler(RequestRepository requestRepository,
 
   private CompletableFuture<Result<ScheduledNotice>> handleRequestNotice(ScheduledNotice notice) {
     return requestRepository.getById(notice.getRequestId())
-      .thenApply(r -> r.map(RequestAndRelatedRecords::new))
-      .thenCompose(r -> r.after(records -> sendNotice(records, notice)))
-      .thenCompose(r -> r.after(records -> updateNotice(records, notice)));
+      .thenCompose(r -> r.afterWhen(
+        request -> noticeShouldBeDeleted(request, notice),
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java:72]:
```suggestion
        request -> noticeIsIrrelevant(request, notice),
```

Notice should also be deleted (after sending) if it is relevant, but not recurring. So the method name `noticeShouldBeDeleted` can be a bit misleading.

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**